k9mail/src/main/java/com/fsck/k9/controller/MessageDownloader.java

```diff
@@ -0,0 +1,713 @@
+package com.fsck.k9.controller;
+
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Date;
+import java.util.EnumSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import android.content.Context;
+
+import com.fsck.k9.Account;
+import com.fsck.k9.AccountStats;
+import com.fsck.k9.K9;
+import com.fsck.k9.Preferences;
+import com.fsck.k9.activity.MessageReference;
+import com.fsck.k9.helper.Contacts;
+import com.fsck.k9.mail.BodyFactory;
+import com.fsck.k9.mail.DefaultBodyFactory;
+import com.fsck.k9.mail.FetchProfile;
+import com.fsck.k9.mail.Flag;
+import com.fsck.k9.mail.Folder;
+import com.fsck.k9.mail.Message;
+import com.fsck.k9.mail.MessageRetrievalListener;
+import com.fsck.k9.mail.MessagingException;
+import com.fsck.k9.mail.Part;
+import com.fsck.k9.mail.internet.MessageExtractor;
+import com.fsck.k9.mailstore.LocalFolder;
+import com.fsck.k9.mailstore.LocalMessage;
+import com.fsck.k9.mailstore.MessageRemovalListener;
+import com.fsck.k9.notification.NotificationController;
+import timber.log.Timber;
+
+import static com.fsck.k9.controller.SyncUtils.modeMismatch;
+
+
+class MessageDownloader {
```

Comment:
I added a flag to control fetching of flags for new messages (used in the QRESYNC case, when we get the flags in the SELECT-QRESYNC response). Other than that, everything is the same.